### ISSUE TITLE
libcurl redirect url and timeouts

### ISSUE THREAD

**Issue by pps83** [curl/curl#2556]
### I did this

I use libcurl multi interface. 
I send an http request to server `A`. Server redirects it to server `B` (using `302`). When libcurl follows redirect and if request to B times out then on libcurl side I get notification that the request completed and `statusCode` is `302` (from previous redirect) and if I try to call `curl_easy_getinfo(CURLINFO_REDIRECT_URL)` I get empty redirect URL. I do not see anything that would instead tell me that the request timed out.

### I expected the following

curl seems to be following its API. E.g. `CURLINFO_RESPONSE_CODE` returns last received `statusCode` and `CURLINFO_REDIRECT_URL` has this:

```
Pass a pointer to a char pointer to receive the URL a redirect would take you to if you would
enable CURLOPT_FOLLOWLOCATION. This can come very handy if you think using the built-in
libcurl redirect logic isn't good enough for you but you would still prefer to avoid implementing
all the magic of figuring out the new URL.

This URL is also set if the CURLOPT_MAXREDIRS limit prevented a redirect to happen (since 7.54.1).
```

1. I want to know that request actually timed out.
2. I don't see a reason why `CURLINFO_REDIRECT_URL` wouldn't return redirected to location regardless of `CURLOPT_FOLLOWLOCATION`. Basically, when I get all the headers (when I get callback for `\r\n\r\n`) it should already be able to return location.
3. it would be nice to have a clear callback to know when a new request (direct) started. So that it would be clear what request actually failed/timed out. Previous that was successful, or the new redirect.

### curl/libcurl version

`#define LIBCURL_VERSION "7.59.0"`

**Comment by pps83** [comment]
for 2. I guess there is a way to figure it out now without having callback (there is an option for ACTUAL_URL or something like that), and there is an option for number of redirects taken. Also, header callback could be used to see `Location` header. If there was a redirect callback default redirect handling would be default callback that follows redirects util max redirects reached. User supplied callback would allow to user code to tell curl to follow given location or not. This is also somewhat in line with msxml http com object (many projects are forced to use it, like xbox games where you have to use it for http requests afaik).

**Comment by pps83** [comment]
1 - as I wrote, there is way to figure it out without a callback. I was also referring to `CURLINFO_EFFECTIVE_URL`:
> there is an option for ACTUAL_URL or something like that)

2. I do use manual redirect in my code, but when i don't use them I do have that issue. I use manual redirect only in some specific cases. But in any case: 
>I don't see a reason why CURLINFO_REDIRECT_URL wouldn't return redirected to location regardless of CURLOPT_FOLLOWLOCATION

> 3 - "there is no way to get result from easy handle alone. The only way it seems is to use CURLMsg." that's an excellent contradiction in there. That is the way to get the result from the easy handle. Why is that a problem?

I don't see where you see contradiction :) What I'm saying is this: when request completes for all kinds of stuff I can query CURL easy handle. Eg to get redirect url, status code etc etc I can get them using `curl_easy_getinfo`. BUT, if I have *ONLY* CURL handle, then I won't be able to get the result of the completed request form the CURL handle alone using curl_easy_getinfo. In debugger I clearly see where the result is (it's obviously part of CURL handle), but no way to retrieve it from given handle using public api.

**Comment by bagder** [comment]
> I don't see a reason why CURLINFO_REDIRECT_URL wouldn't return redirected to location regardless of CURLOPT_FOLLOWLOCATION

Because I made it like that? I wouldn't be totally against changing it to get set unconditionally.

> I have ONLY CURL handle, then I won't be able to get the result of the completed request form the CURL handle alone using curl_easy_getinfo.

I really don't understand this argument. What handle are you talking about here? You added all the easy handles to the multi handle, they're all still there and libcurl tells you which handles that complete. Why can't you use those handles to extract information?

**Comment by pps83** [comment]
> I really don't understand this argument. What handle are you talking about here? You added all the easy handles to the multi handle, they're all still there and libcurl tells you which handles that complete. Why can't you use those handles to extract information?

Yes, that's exactly what I do :) I'll rephrase it differently, so it would be clear what I'm talking about. I think there has to be CURLOPT_RESULT that could retrieve `result` from CURL easy handle. This way identical code could retrieve result regardless if it's multi or easy interface.

**Comment by bagder** [comment]
You mean like this?
~~~c
    msg = curl_multi_info_read(multi, &queued);
    if(msg && (msg->msg == CURLMSG_DONE)) {
        CURL *e = msg->easy_handle;
        CURLcode result = msg->data.result;
        curl_easy_getinfo(msg->easy_handle, CURLINFO_..., ...);
...

**Comment by bagder** [comment]
I can't say that I agree with you that curl_easy_getinfo() needs such an option, but I will agree with you that there is no such one today.
